core/client/fs/src/main/java/alluxio/client/block/AlluxioBlockStore.java

```diff
@@ -131,28 +133,62 @@ public BlockInfo getInfo(long blockId) throws IOException {
+   * determining the data source and type of data source. The latest BlockInfo will be fetched
+   * from the master to ensure the locations are up to date.
+   *
+   * @param blockId the id of the block to read
+   * @param options the options associated with the read request
+   * @param failedWorkers the map of workers address to most recent failure time
+   * @return a stream which reads from the beginning of the block
+   */
+  public BlockInStream getInStream(long blockId, InStreamOptions options,
+      Map<WorkerNetAddress, Long> failedWorkers) throws IOException {
     // Get the latest block info from master
     BlockInfo info;
     try (CloseableResource<BlockMasterClient> masterClientResource =
              mContext.acquireBlockMasterClientResource()) {
       info = masterClientResource.get().getBlockInfo(blockId);
     }
-    List<BlockLocation> locations = info.getLocations();
-    if (locations.isEmpty() && !options.getStatus().isPersisted()) {
+    List<BlockLocation> allLocations = info.getLocations();
+    boolean isPersisted = options.getStatus().isPersisted();
+    if (allLocations.isEmpty() && !isPersisted) {
       throw new NotFoundException(ExceptionMessage.BLOCK_UNAVAILABLE.getMessage(info.getBlockId()));
     }
     // Determine the data source and the type of data source
     // TODO(calvin): Consider containing these two variables in one object
     BlockInStreamSource dataSourceType = null;
     WorkerNetAddress dataSource = null;
-    if (locations.isEmpty()) { // Data will be read from UFS
-      dataSourceType = BlockInStreamSource.UFS;
-      BlockLocationPolicy policy =
-          Preconditions.checkNotNull(options.getOptions().getUfsReadLocationPolicy(),
-              PreconditionMessage.UFS_READ_LOCATION_POLICY_UNSPECIFIED);
-      GetWorkerOptions getWorkerOptions = GetWorkerOptions.defaults().setBlockId(info.getBlockId())
-          .setBlockSize(info.getLength()).setBlockWorkerInfos(getEligibleWorkers());
-      dataSource = policy.getWorker(getWorkerOptions);
+    List<BlockLocation> locations = allLocations.stream().filter(
+        location -> !failedWorkers.containsKey(location.getWorkerAddress()))
+        .collect(Collectors.toList());
+    if (locations.isEmpty()) {
+      List<BlockWorkerInfo> allWorkers = getEligibleWorkers();
+      List<BlockWorkerInfo> availableWorkers = allWorkers.stream()
+          .filter(location -> !failedWorkers.containsKey(location.getNetAddress()))
+          .collect(Collectors.toList());
+      if (availableWorkers.isEmpty()) {
+        // we run out of workers to read from, fallback to the least recently failed worker
+        dataSource = (isPersisted
+                ? allWorkers.stream().map(BlockWorkerInfo::getNetAddress)
+                : allLocations.stream().map(BlockLocation::getWorkerAddress))
+                .min((x, y) -> Long.compare(failedWorkers.get(x), failedWorkers.get(y))).get();
+        dataSourceType = getDataSourceType(dataSource, info);
```

Comment:
shouldn't `dataSourceType` here always be `BlockInStreamSource.UFS`, because `locations.isEmpty()`?

Reply:
In this case a new dataSource is selected due to fallback logic and it can be a UFS data source or local/remote depends on which worker it falls back to.

Reply:
even when `locations.isEmpty()`, can we still fallback to local/remote worker?